ta/pkcs11/src/token_capabilities.c

```diff
@@ -0,0 +1,240 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2017-2020, Linaro Limited
+ */
+
+#include <assert.h>
+#include <pkcs11_ta.h>
+#include <string.h>
+#include <util.h>
+
+#include "pkcs11_helpers.h"
+#include "token_capabilities.h"
+
+/*
+ * Definition of supported processings for a PKCS#11 mechanisms
+ * @id: Mechanism ID
+ * @flags: Valid PKCS11_CKFM_* for a mechanism as per PKCS#11
+ * @one_shot: true of mechanism can be used for a one-short processing
+ * @string: Helper string of the mechanism ID for debug purpose
+ */
+struct pkcs11_mechachism_modes {
+	uint32_t id;
+	uint32_t flags;
+	bool one_shot;
+#if CFG_TEE_TA_LOG_LEVEL > 0
+	const char *string;
+#endif
+};
+
+#if CFG_TEE_TA_LOG_LEVEL > 0
+#define MECHANISM(_label, _flags, _one_shot)	\
+	{					\
+		.id = _label,			\
+		.one_shot = (_one_shot),	\
+		.flags = (_flags),		\
+		.string = #_label,		\
+	}
+#else
+#define MECHANISM(_label, _flags, _one_shot)	\
+	{					\
+		.id = _label,			\
+		.one_shot = (_one_shot),	\
+		.flags = (_flags),		\
+	}
+#endif
+
+#define MECHA(_label, _dig, _enc, _dec, _sig, _ver,		\
+	       _sr, _vr, _der, _wra, _unw, _gen, _gpa, _1s)	\
+	MECHANISM(PKCS11_CKM_ ## _label,				\
+		  (_enc ? PKCS11_CKFM_ENCRYPT : 0) |		\
+		  (_dec ? PKCS11_CKFM_DECRYPT : 0) |		\
+		  (_dig ? PKCS11_CKFM_DIGEST : 0) |		\
+		  (_sig ? PKCS11_CKFM_SIGN : 0) |			\
+		  (_sr ? PKCS11_CKFM_SIGN_RECOVER : 0) |		\
+		  (_ver ? PKCS11_CKFM_VERIFY : 0) |		\
+		  (_vr ? PKCS11_CKFM_VERIFY_RECOVER : 0) |	\
+		  (_gen ? PKCS11_CKFM_GENERATE : 0) |		\
+		  (_gpa ? PKCS11_CKFM_GENERATE_KEY_PAIR : 0) |	\
+		  (_wra ? PKCS11_CKFM_WRAP : 0) |		\
+		  (_unw ? PKCS11_CKFM_UNWRAP : 0) |		\
+		  (_der ? PKCS11_CKFM_DERIVE : 0),		\
+		  _1s)
+
+static const struct pkcs11_mechachism_modes pkcs11_modes[] = {
+	/*
+	 * PKCS#11 directives on mechanism support for the several processing
+	 * modes.
+	 *				1: One shot processing supported ---.
+	 *				Gp: Generate secret pair --------.  |
+	 *				Ge: Generate secret value ----.  |  |
+	 *				Wr|Uw: Wrap/Unwrap -------.   |  |  |
+	 *				Dr: Derive ----------.    |   |  |  |
+	 *		Sr|Vr: SignRecover/VerifyRecov --.   |    |   |  |  |
+	 *		Si|Ve: Sign/Verify --------.     |   |    |   |  |  |
+	 *		En|De: Encrypt/Decrypt     |     |   |    |   |  |  |
+	 *		Di: Digest -----.    |     |     |   |    |   |  |  |
+	 *				|   / \   / \   / \  |   / \  |  |  |
+	 * Mechanism			Di|En|De|Si|Ve|Sr|Vr|Dr|Wr|Uw|Ge|Gp|1
+	 */
+	MECHA(AES_ECB,			0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0),
```

Comment:
It is a bit hard to match att the flags, I find this:
``` C
MECHANISM(PKCS11_CMD_AES_ECB,
          PKCS11_CKFM_ENCRYPT | PKCS11_CKFM_DECRYPT |
          PKCS11_CKFM_DERIVE | PKCS11_CKFM_WRAP | PKCS11_CKFM_UNWRAP),
```
rather readable with the advantage of being not so complicated.

Reply:
Note that this table only defines the PKCS#11 specification rules. Another table defines what
we support in the TA, see line 188.

Considering a quite large number of mechanism, I prefer my proposal where things match in single line for the PKCS#11 spec part, for example from my dev branch: [PKCS#11 spec for mechanism processing](https://github.com/etienne-lms/optee_os/blob/sks-to-pkcs11/ta/pkcs11/src/pkcs11_attributes.c#L68-L141) and [list of mechanism/processing supported by the TA](https://github.com/etienne-lms/optee_os/blob/sks-to-pkcs11/ta/pkcs11/src/pkcs11_token.c#L610).

Reply:
Well, I'll in an explicit flag based way for the PKCS11 spec part and see how it looks like.

Reply:
See temporary added [hack commit](https://github.com/OP-TEE/optee_os/pull/3667/commits/1b6d3183b4b18aecc7182297611be2dce1f9efc3#diff-80dfbd2ecfb1e100eae626803e0f88e6R68) for what it could look like for a larger list of mechanisms.